### PR TITLE
Build WeWorkRemotely scraper closes #1

### LINKED ISSUE #1: We Work Remotely Scraper
Size: 3

We Work Remote features remote jobs postings. They often have job postings that Stack Overflow doesn't have, so we need a scraper to pull in the job postings.

Acceptance Criteria:
- Job Fetch Service pulls in jobs from We Work Remote's RSS Feed
- Service is tested
- Service has an associated rake task to run it.

Resources:
https://weworkremotely.com/categories/2-programming/jobs.rss

### PR DESCRIPTION
**Pull request by patwey**
Adds WeWorkRemotely service and rake task to run it

### CODE REVIEW

File: app/services/we_work_remotely.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  def self.scrub_description(description)
+    description.gsub(NEWLINES_OR_TABS, " ")
+               .gsub(ESCAPED_QUOTES, "'")
+               .split.join(" ")
+               .strip
+  end
+
+  def self.pull_feed
+    url = BASE_URI
+    Feedjira::Feed.fetch_and_parse url
```
Comment (rrgayhart) [inline app/services/we_work_remotely.rb]: You could probably get away with doing this in one line without sacrificing readability

File: app/services/we_work_remotely.rb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  BASE_URI = "https://weworkremotely.com/categories/2-programming/jobs.rss"
+
+  DIVS_OR_LIS        = /<\/div>|<li>/
+  NONBREAKING_SPACES = /&nbsp;/
+  NEWLINES_OR_TABS   = /\n+|\t+/
+  ESCAPED_QUOTES     = "\""
+
+  def self.scrape
+    technologies = Technology.pluck(:name)
+    feed = self.pull_feed
```
Comment (rrgayhart) [inline app/services/we_work_remotely.rb]: It would be nice to have a test for what happens if WeWorkRemote goes down when we run the scraper

**Comment by worace** [comment]
@patwey gave this a quick look and i think it is pretty solid. appreciate the tests you added and general clarity of the code.

I'm a little uncertain about the decision to make everything in this new class and the JobFetcher base singleton methods rather than just instantiating new instances of the services when needed, but given that that is the pattern in place i appreciate that you stuck to it consistently.

Running it locally i pulled in these companies which seems good:

```
"Forward Action", "Devschool", "Toptal", "Mycelium Swish", "Seat Engine Ticketing", "Drip", "Help Scout", "Telemetry", "Ezhome", "Follow Up Boss", "Gear Stream", "Truetester", "Basecamp Networks", "Merit Pages, Inc.: RoR Developer at a Not-Bullshit Company (in Albany, NY not Silicon Valley D", "Barrel", "BigWig", "Ruzuku", "Clevertech", "PrintFriendly.com"]
```

It does look like this one: `"Merit Pages, Inc.: RoR Developer at a Not-Bullshit Company (in Albany, NY not Silicon Valley D"` caused a little hiccup b/c of the "Inc." -- not sure how serious that edge case is or not.

Ran tests after merging into master and all passed so that looks good. left a few other small comments inline but otherwise :+1: from me

File: app/services/we_work_remotely.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  end
+
+  def self.format_entries(entries, technologies)
+    entries.map do |entry|
+      formatted_entry = self.format_entry(entry, technologies)
+      self.create_records(formatted_entry)
+    end
+  end
+
+  def self.format_entry(entry, technologies)
```
Comment (rrgayhart) [inline app/services/we_work_remotely.rb]: We're passing technologies around a lot here - might be better to define it where it's needed or even move the definition of technologies into the JobFetcher class

File: app/services/we_work_remotely.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    summary = strip_summary(entry.summary)
+    description = self.pull_description(summary)
+
+    { job: {
+        title: entry.title,
+        url: entry.url,
+        location: self.pull_location(summary),
+        raw_technologies: self.pull_technologies(description, technologies),
+        description: description,
+        remote: true,
```
Comment (rrgayhart) [inline app/services/we_work_remotely.rb]: Nice!

File: spec/services/weworkremotely_spec.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+  describe '.pull_company_name' do
+    it 'pulls a company from a title' do
+      title = "Basecamp Networks: Computer Vision Engineer"
+      company_name = service.pull_company_name(title)
+
+      expect(company_name).to eq('Basecamp Networks')
+    end
+
+    it 'handles unmatching pattern' do
```
Comment (worace) [inline spec/services/weworkremotely_spec.rb]: for a test like this i sometimes like to include text describing how it handles it  (In this case i guess something like "handles unmatching pattern by leaving name blank"). Test itself is good but this helps a little bit with the "documentation" side of things

**Comment by worace** [comment]
The only other big question around something like this is whether to include a real "integration" test against some canned network data. for your purposes it would require throwing in something like a VCR cassette or stubbing against this FeedJira thing.

That kind of testing is tedious but unfortunately is the only way to really test against all those weird edge cases that come up. Up to you guys about whether you think it's important, but from the perspective of completion that's probably the main thing that's missing here.

File: app/services/we_work_remotely.rb [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  end
+
+  def self.pull_location(summary)
+    regex = /Headquarters: (.*)\s* URL/
+    regex.match(summary)[1] rescue ''
+  end
+
+  def self.pull_technologies(description, technologies)
+    technologies.select do |tech|
+      regex = /\b#{tech}\b/i
```
Comment (rrgayhart) [inline app/services/we_work_remotely.rb]: This is a great way to do this! May I also suggest: 
```
  Regexp.new(technologies.join("|"))
```

**Comment by rrgayhart** [comment]
Pat - you bring honor to your family! Great work!

File: app/services/we_work_remotely.rb (hunk 3 above)
Comment (patwey) [inline app/services/we_work_remotely.rb]: Good catch! Who do I think this service is, the Peyton Manning of technologies?... ... ...

I'll move it to the JobFetcher class

File: app/services/we_work_remotely.rb (hunk 6 above)
Comment (patwey) [inline app/services/we_work_remotely.rb]: @rrgayhart I'm having trouble implementing this. Maybe we could pair on it tomorrow?

File: app/services/we_work_remotely.rb (hunk 1 above)
Comment (patwey) [inline app/services/we_work_remotely.rb]: Done!

File: spec/services/weworkremotely_spec.rb (hunk 5 above)
Comment (patwey) [inline spec/services/weworkremotely_spec.rb]: Good call sir!

**Comment by patwey** [comment]
@worace Thanks for looking it over! 

Regarding the `"Merit Pages, Inc.: RoR Developer at a Not-Bullshit Company (in Albany, NY not Silicon Valley D"` title: I just looked at the feed, and that's what they specified within the `<title>` tags of the job posting.  I could add something to try to parse the titles, but I don't have any ideas on how to decide what to include/exclude from what they've listed.

I'd like to add an integration test, but I don't know if I'll be able to get that done by tomorrow.

**Comment by rrgayhart** [comment]
@patwey - Can you add an issue to add 'integration' level testing to this service? That way someone can snag it!

**Comment by rrgayhart** [comment]
:+1: :ship: it!
